walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,132 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long duration = request.getDuration();
+
+    if (attendees.isEmpty() && optionalAttendees.isEmpty()){
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }else 
+    if (duration > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+    }
+
+    List<TimeRange> takenTime = new ArrayList<TimeRange>();
+    List<TimeRange> optionalBusy = new ArrayList<TimeRange>();
+    setTakenTimes(events, takenTime, optionalBusy, attendees, optionalAttendees);
+
+    List<TimeRange> mandatoryFree = new ArrayList<TimeRange>();
+    mandatoryFree.add(TimeRange.WHOLE_DAY);
+    List<TimeRange> optionalFree = new ArrayList<TimeRange>();
+    setFreeTimes(mandatoryFree, takenTime, duration);
+
+    if(takenTime.isEmpty()){
+      optionalFree.add(TimeRange.WHOLE_DAY);
+      setFreeTimes(optionalFree, optionalBusy, duration);
+      return optionalFree;
+    }else {
+      for(TimeRange t: mandatoryFree){
+        checkOptional(optionalFree, optionalBusy, t);
+      }
+    }
```

Comment:
You have "return" in first "if" block, which makes redundant to use "else" block for the "for" loop.

Reply:
Okay sure Tong. I removed the else as suggested.